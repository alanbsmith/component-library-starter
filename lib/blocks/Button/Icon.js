import styled from 'styled-components';
import { rem } from 'polished';
import { applyStyleModifiers } from 'styled-components-modifiers';

import { Icon as BasicIcon } from '../../elements';

const modifierConfig = {
  left: (props) => `
    padding-right: ${rem(props.theme.dimensions.baseGrid)};
  `,
  right: (props) => `
    padding-left: ${rem(props.theme.dimensions.baseGrid)};
  `,
};

const Icon = styled(BasicIcon)`
  font-size: 24px;
  ${applyStyleModifiers(modifierConfig)}
`;

export default Icon;
