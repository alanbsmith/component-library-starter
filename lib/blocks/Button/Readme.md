```jsx
<div style={{ display: 'flex', flexDirection: 'row', flexWrap: 'wrap' }}>
  <div style={{ flex: '1', padding: '5px' }}>
    <Button>
      <Button.Icon name="home" modifiers={['left', 'chrome000']} />
      <Button.Text>
        action
      </Button.Text>
    </Button>
  </div>
  <div style={{ flex: '1', padding: '5px' }}>
    <Button modifiers={['bgPink', 'hoverBgPink']}>
      <Button.Icon name="home" modifiers={['left', 'chrome000']} />
      <Button.Text>
        action
      </Button.Text>
    </Button>
  </div>
  <div style={{ flex: '1', padding: '5px' }}>
    <Button modifiers={['bgBlue', 'hoverBgBlue']}>
      <Button.Icon name="home" modifiers={['left', 'chrome000']} />
      <Button.Text>
        action
      </Button.Text>
    </Button>
  </div>
  <div style={{ flex: '1', padding: '5px' }}>
    <Button modifiers={['bgPurple', 'hoverBgPurple']}>
      <Button.Icon name="home" modifiers={['left', 'chrome000']} />
      <Button.Text>
        action
      </Button.Text>
    </Button>
  </div>
  <div style={{ flex: '1', padding: '5px' }}>
    <Button modifiers={['bgGreen', 'hoverBgGreen']}>
      <Button.Icon name="home" modifiers={['left', 'chrome000']} />
      <Button.Text>
        action
      </Button.Text>
    </Button>
  </div>
  <div style={{ flex: '1', padding: '5px' }}>
    <Button modifiers={['bgOrange', 'hoverBgOrange']}>
      <Button.Icon name="home" modifiers={['left', 'chrome000']} />
      <Button.Text>
        action
      </Button.Text>
    </Button>
  </div>
</div>
```
