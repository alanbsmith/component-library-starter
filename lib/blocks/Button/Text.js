import styled from 'styled-components';

import { Text as BaseText } from '../../elements';

const Text = styled(BaseText)`
  font-weight: ${props => props.theme.fontWeights.light};
  text-transform: uppercase;
`;

export default Text;
