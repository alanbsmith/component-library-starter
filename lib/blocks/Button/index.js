import styled from "styled-components";
import { lighten, rem } from "polished";
import { applyStyleModifiers } from "styled-components-modifiers";

import Icon from "./Icon";
import Text from "./Text";

import { bgColors, hoverBgColors } from "../../modifiers";

import gridScale from "../../utils/gridScale";

const customModifiers = {
  fullWidth: () => `flex: 1`
};

const modifierConfig = {
  ...bgColors,
  ...hoverBgColors,
  ...customModifiers
};

const Button = styled.button`
  align-items: flex-end;
  background-color: ${props => props.theme.colors.chrome500};
  border-radius: ${props => props.theme.dimensions.borderRadius};
  border: solid 1px transparent;
  color: ${props => props.theme.colors.chrome000};
  cursor: pointer;
  display: flex;
  font-size: ${props => rem(props.theme.dimensionsFontSize)};
  justify-content: center;
  padding: ${props => `${gridScale(props, 1)} ${gridScale(props, 3)}`};
  outline: none;
  transition: all 200ms ease;

  &:hover,
  &:focus {
    background-color: ${props => lighten(0.05, props.theme.colors.chrome500)};
    box-shadow: 0 8px 8px 0 ${props => props.theme.colors.shadow};
  }

  ${applyStyleModifiers(modifierConfig)};
`;

Button.Icon = Icon;
Button.Text = Text;

/** @component */
export default Button;
