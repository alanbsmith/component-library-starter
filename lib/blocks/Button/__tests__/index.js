import React from 'react';
import Button from '../index';
import 'jest-styled-components';

import { renderWithTheme } from '__tests__/helpers';

describe('Button Block', () => {
  it('renders correctly', () => {
    const tree = renderWithTheme(
      <Button>
        <Button.Icon
          modifiers={['left', 'chrome000']}
          name="home"
        />
        <Button.Text>
          action
        </Button.Text>
      </Button>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('applies a bgColor modifier', () => {
    const tree = renderWithTheme(
      <Button modifiers={['bgPink']}>
        <Button.Icon
          modifiers={['left', 'chrome000']}
          name="home"
        />
        <Button.Text>
          action
        </Button.Text>
      </Button>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('applies a hoverBgColor modifier', () => {
    const tree = renderWithTheme(
      <Button modifiers={['hoverBgGreen']}>
        <Button.Icon
          modifiers={['left', 'chrome000']}
          name="home"
        />
        <Button.Text>
          action
        </Button.Text>
      </Button>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('applies the fullWidth modifier', () => {
    const tree = renderWithTheme(
      <Button modifiers={['fullWidth']}>
        <Button.Icon
          modifiers={['left', 'chrome000']}
          name="home"
        />
        <Button.Text>
          action
        </Button.Text>
      </Button>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
