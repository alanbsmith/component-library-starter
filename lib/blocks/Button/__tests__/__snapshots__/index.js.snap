// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Block applies a bgColor modifier 1`] = `
.c0 {
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  background-color: #9E9E9E;
  border-radius: 2px;
  border: solid 1px transparent;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: NaNrem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.5rem 1.5rem;
  outline: none;
  -webkit-transition: all 200ms ease;
  transition: all 200ms ease;
  background-color: #E91E63;
}

.c0:hover,
.c0:focus {
  background-color: #ababab;
  box-shadow: 0 8px 8px 0 rgba(117,117,117,0.4);
}

.c1 {
  font-size: 24px;
  padding-right: 0.5rem;
}

.c2 {
  color: #424242;
  font-size: 1.25rem;
  color: #FFFFFF;
}

.c3 {
  font-weight: 300;
  text-transform: uppercase;
}

.c4 {
  font-size: 1rem;
  margin: 0;
}

<button
  className="c0"
>
  <span
    aria-hidden={true}
    className="fa fa-home c1 c2"
    modifiers={
      Array [
        "left",
        "chrome000",
      ]
    }
  />
  <span
    className="c3 c4"
  >
    action
  </span>
</button>
`;

exports[`Button Block applies a hoverBgColor modifier 1`] = `
.c0 {
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  background-color: #9E9E9E;
  border-radius: 2px;
  border: solid 1px transparent;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: NaNrem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.5rem 1.5rem;
  outline: none;
  -webkit-transition: all 200ms ease;
  transition: all 200ms ease;
}

.c0:hover,
.c0:focus {
  background-color: #ababab;
  box-shadow: 0 8px 8px 0 rgba(117,117,117,0.4);
}

.c0:hover,
.c0:focus {
  background-color: #5cb860;
}

.c1 {
  font-size: 24px;
  padding-right: 0.5rem;
}

.c2 {
  color: #424242;
  font-size: 1.25rem;
  color: #FFFFFF;
}

.c3 {
  font-weight: 300;
  text-transform: uppercase;
}

.c4 {
  font-size: 1rem;
  margin: 0;
}

<button
  className="c0"
>
  <span
    aria-hidden={true}
    className="fa fa-home c1 c2"
    modifiers={
      Array [
        "left",
        "chrome000",
      ]
    }
  />
  <span
    className="c3 c4"
  >
    action
  </span>
</button>
`;

exports[`Button Block applies the fullWidth modifier 1`] = `
.c0 {
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  background-color: #9E9E9E;
  border-radius: 2px;
  border: solid 1px transparent;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: NaNrem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.5rem 1.5rem;
  outline: none;
  -webkit-transition: all 200ms ease;
  transition: all 200ms ease;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0:hover,
.c0:focus {
  background-color: #ababab;
  box-shadow: 0 8px 8px 0 rgba(117,117,117,0.4);
}

.c1 {
  font-size: 24px;
  padding-right: 0.5rem;
}

.c2 {
  color: #424242;
  font-size: 1.25rem;
  color: #FFFFFF;
}

.c3 {
  font-weight: 300;
  text-transform: uppercase;
}

.c4 {
  font-size: 1rem;
  margin: 0;
}

<button
  className="c0"
>
  <span
    aria-hidden={true}
    className="fa fa-home c1 c2"
    modifiers={
      Array [
        "left",
        "chrome000",
      ]
    }
  />
  <span
    className="c3 c4"
  >
    action
  </span>
</button>
`;

exports[`Button Block renders correctly 1`] = `
.c0 {
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  background-color: #9E9E9E;
  border-radius: 2px;
  border: solid 1px transparent;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: NaNrem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.5rem 1.5rem;
  outline: none;
  -webkit-transition: all 200ms ease;
  transition: all 200ms ease;
}

.c0:hover,
.c0:focus {
  background-color: #ababab;
  box-shadow: 0 8px 8px 0 rgba(117,117,117,0.4);
}

.c1 {
  font-size: 24px;
  padding-right: 0.5rem;
}

.c2 {
  color: #424242;
  font-size: 1.25rem;
  color: #FFFFFF;
}

.c3 {
  font-weight: 300;
  text-transform: uppercase;
}

.c4 {
  font-size: 1rem;
  margin: 0;
}

<button
  className="c0"
>
  <span
    aria-hidden={true}
    className="fa fa-home c1 c2"
    modifiers={
      Array [
        "left",
        "chrome000",
      ]
    }
  />
  <span
    className="c3 c4"
  >
    action
  </span>
</button>
`;
